aspects/schema_test.go

```diff
@@ -1909,6 +1925,178 @@ func (*schemaSuite) TestUnexpectedTypes(c *C) {
 
 		input := []byte(fmt.Sprintf(`{"foo": %v}`, tc.testValue))
 		err = schema.Validate(input)
-		c.Assert(err, ErrorMatches, fmt.Sprintf(`cannot accept element in "foo": expected %s type but got %T`, tc.expectedType, tc.testValue))
+		c.Assert(err, ErrorMatches, fmt.Sprintf(`cannot accept element in "foo": expected %s type but value was %T`, tc.expectedType, tc.testValue))
+	}
+}
+
+func (*schemaSuite) TestAlternativeTypesHappy(c *C) {
+	schemaStr := []byte(`{
+	"schema": {
+		"foo": ["string", "int"]
+	}
+}`)
+
+	schema, err := aspects.ParseSchema(schemaStr)
+	c.Assert(err, IsNil)
+
+	for _, val := range []interface{}{`"one"`, `1`} {
+		input := []byte(fmt.Sprintf(`{"foo":%s}`, val))
+		err = schema.Validate(input)
+		c.Assert(err, IsNil)
+	}
+}
+
+func (*schemaSuite) TestAlternativeTypesFail(c *C) {
+	schemaStr := []byte(`{
+	"schema": {
+		"foo": ["string", "int"]
+	}
+}`)
+
+	schema, err := aspects.ParseSchema(schemaStr)
+	c.Assert(err, IsNil)
+
+	for _, val := range []interface{}{"1.1", "true", `{"bar": 1}`, `[1, 2]`} {
+		input := []byte(fmt.Sprintf(`{"foo":%s}`, val))
+		err = schema.Validate(input)
+		c.Assert(err, ErrorMatches, `cannot accept element in "foo": no matching schema:
+	expected string .*
+	expected int .*`)
+	}
+}
+
+func (*schemaSuite) TestAlternativeTypesWithConstraintsHappy(c *C) {
+	schemaStr := []byte(`{
+	"schema": {
+		"foo": [
+			{
+				"type": "int",
+				"min": 0
+			},
+			{
+				"type": "string",
+				"pattern": "[bB]ar"
+			}
+		]
+	}
+}`)
+	schema, err := aspects.ParseSchema(schemaStr)
+	c.Assert(err, IsNil)
+
+	for _, val := range []interface{}{"3", "0", `"Bar"`, `"bar"`} {
+		input := []byte(fmt.Sprintf(`{"foo":%s}`, val))
+		err = schema.Validate(input)
+		c.Assert(err, IsNil)
+	}
+}
+
+func (*schemaSuite) TestAlternativeTypesWithConstraintsFail(c *C) {
+	schemaStr := []byte(`{
+	"schema": {
+		"foo": [
+			{
+				"type": "int",
+				"min": 0
+			},
+			{
+				"type": "string",
+				"pattern": "[bB]ar"
+			}
+		]
+	}
+}`)
+	schema, err := aspects.ParseSchema(schemaStr)
+	c.Assert(err, IsNil)
+
+	err = schema.Validate([]byte(`{"foo":-1}`))
+	c.Assert(err, ErrorMatches, `cannot accept element in "foo": no matching schema:
+	-1 is less than the allowed minimum 0
+	expected string type but value was number`)
+
+	err = schema.Validate([]byte(`{"foo":"bAR"}`))
+	c.Assert(err, ErrorMatches, `cannot accept element in "foo": no matching schema:
+	expected int type but value was string
+	expected string matching \[bB\]ar but value was "bAR"`)
+}
```

Comment:
this is better, but:

expected
or expected ...

would still read better?   Also it would be nice to have only one/the relevant "but" bit, we might want to leave a TODO about that for now, but my idea to achieve that would be for ValidationError to carry:

expected type:    string,int,map...
actual type: ...
something optional about unfulfilled constraints beyond the type

here we have two errors:
1. expected int, got string
2. expected string,  got string,  constraint "matching \[bB\]ar",  'but value was "bAR"'

from these info we can omit printing the but part of anything that is expected by another error

expected int
or expected string "matching \[bB\]ar"
but value was "bAR"

you would have to play with the idea a bit, that's why I'm suggesting to this as a follow up

Reply:
I added the "or" and a TODO to work on this as a follow-up